File: web/src/api/nl2sql.ts
/**
 * NL2SQL 自然语言转 SQL API
 *
 * API 规范:
 * - v1 版本使用统一的 ApiResponse 格式
 * - 保留旧版 API 以确保向后兼容
 */

import client from './client';
import type { ApiResponse } from './types';

// 重新导出统一类型
export type { ApiResponse };

// ============================================================
// 类型定义
// ============================================================

export interface NL2SQLQueryRequest {
  query: string;
  database?: string;
  context?: string;
}

export interface NL2SQLQueryResponse {
  sql: string;
  generated_sql?: string;
  explanation?: string;
  confidence?: number;
  tables?: string[];
  columns?: string[];
  rows?: Array<Array<string | number | boolean | null>>;
  row_count?: number;
  execution_time_ms?: number;
  success?: boolean;
}

export interface TableInfo {
  name: string;
  table_name?: string;
  schema?: string;
  database?: string;
  comment?: string;
  columns?: Array<{
    name: string;
    type: string;
    data_type?: string;
    description?: string;
    comment?: string;
  }>;
}

export interface ExplainRequest {
  sql: string;
  database?: string;
}

export interface ExplainResponse {
  sql: string;
  explanation: string;
  steps?: string[];
}

// ============================================================
// v1 版本 API（推荐使用，统一响应格式）
// ============================================================

/** 自然语言查询 */
export async function queryV1(request: NL2SQLQueryRequest): Promise<ApiResponse<NL2SQLQueryResponse>> {
  const resp = await client.post<ApiResponse<NL2SQLQueryResponse>>(
    '/api/proxy/nl2sql/v1/query',
    request
  );
  return resp.data;
}

/** SQL 解释 */
export async function explainV1(params: ExplainRequest): Promise<ApiResponse<ExplainResponse>> {
  const resp = await client.post<ApiResponse<ExplainResponse>>(
    '/api/proxy/nl2sql/v1/explain',
    params
  );
  return resp.data;
}

/** 获取表列表 */
export async function getTablesV1(): Promise<ApiResponse<TableInfo[]>> {
  const resp = await client.get<ApiResponse<TableInfo[]>>(
    '/api/proxy/nl2sql/v1/tables'
  );
  return resp.data;
}

// ============================================================
// 旧版 API（向后兼容，逐步废弃）
// ============================================================

import { NL2SQLQueryRequest as LegacyQueryRequest, NL2SQLQueryResponse as LegacyQueryResponse, TableInfo as LegacyTableInfo } from '../types';

// 自然语言查询
export const query = async (request: LegacyQueryRequest): Promise<LegacyQueryResponse> => {
  const response = await client.post<LegacyQueryResponse>('/api/proxy/nl2sql/api/nl2sql/query', request);
  return response.data;
};

// SQL 解释
export const explain = async (sql: string, database?: string): Promise<{ sql: string; explanation: string }> => {
  const response = await client.post('/api/proxy/nl2sql/api/nl2sql/explain', { sql, database });
  return response.data;
};

// 获取表列表
export const getTables = async (): Promise<LegacyTableInfo[]> => {
  const response = await client.get<LegacyTableInfo[]>('/api/proxy/nl2sql/api/nl2sql/tables');
  return response.data;
};

File: services/portal/routers/nl2sql.py
"""NL2SQL 代理路由 - 自然语言查询"""

import os
from typing import Optional

import httpx
from fastapi import APIRouter, Depends, Request, Response
from pydantic import BaseModel

from services.common.api_response import ApiResponse, ErrorCode, error, success
from services.common.auth import TokenPayload, get_current_user
from services.portal.config import settings
from services.portal.routers.proxy import proxy_request

# 服务间通信密钥
SERVICE_SECRET = os.environ.get("SERVICE_SECRET", "internal-service-secret-dev-do-not-use-in-prod")

router = APIRouter(prefix="/api/proxy/nl2sql", tags=["NL2SQL"])


# ============================================================
# 请求模型
# ============================================================

class QueryRequest(BaseModel):
    query: str
    database: Optional[str] = None
    context: Optional[str] = None


class ExplainRequest(BaseModel):
    sql: str
    database: Optional[str] = None


# ============================================================
# v1 版本 API（推荐使用，统一响应格式）
# ============================================================

@router.post("/v1/query", response_model=ApiResponse)
async def query_v1(
    request: QueryRequest,
    user: TokenPayload = Depends(get_current_user),
) -> ApiResponse:
    """自然语言查询（v1 API）"""
    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            resp = await client.post(
                f"{settings.NL2SQL_URL}/api/nl2sql/query",
                json=request.model_dump(),
                headers={"X-Service-Secret": SERVICE_SECRET},
            )
            if resp.status_code == 200:
                return success(data=resp.json())
            return error(
                message=f"NL2SQL 服务错误: {resp.status_code}",
                code=ErrorCode.EXTERNAL_SERVICE_ERROR,
            )
    except Exception as e:
        return error(
            message=f"NL2SQL 服务异常: {str(e)}",
            code=ErrorCode.EXTERNAL_SERVICE_ERROR,
        )


@router.post("/v1/explain", response_model=ApiResponse)
async def explain_v1(
    request: ExplainRequest,
    user: TokenPayload = Depends(get_current_user),
) -> ApiResponse:
    """SQL 解释（v1 API）"""
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            resp = await client.post(
                f"{settings.NL2SQL_URL}/api/nl2sql/explain",
                json=request.model_dump(),
                headers={"X-Service-Secret": SERVICE_SECRET},
            )
            if resp.status_code == 200:
                return success(data=resp.json())
            return error(
                message=f"NL2SQL 服务错误: {resp.status_code}",
                code=ErrorCode.EXTERNAL_SERVICE_ERROR,
            )
    except Exception as e:
        return error(
            message=f"NL2SQL 服务异常: {str(e)}",
            code=ErrorCode.EXTERNAL_SERVICE_ERROR,
        )


@router.get("/v1/tables", response_model=ApiResponse)
async def get_tables_v1(
    user: TokenPayload = Depends(get_current_user),
) -> ApiResponse:
    """获取表列表（v1 API）"""
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            resp = await client.get(
                f"{settings.NL2SQL_URL}/api/nl2sql/tables",
                headers={"X-Service-Secret": SERVICE_SECRET},
            )
            if resp.status_code == 200:
                return success(data=resp.json())
            return error(
                message=f"NL2SQL 服务错误: {resp.status_code}",
                code=ErrorCode.EXTERNAL_SERVICE_ERROR,
            )
    except Exception as e:
        return error(
            message=f"NL2SQL 服务异常: {str(e)}",
            code=ErrorCode.EXTERNAL_SERVICE_ERROR,
        )


# ============================================================
# 旧版 API（向后兼容，直接代理）
# ============================================================

@router.api_route("/{path:path}", methods=["GET", "POST", "PUT", "DELETE"])
async def nl2sql_proxy(
    path: str,
    request: Request,
    user: TokenPayload = Depends(get_current_user),
) -> Response:
    """代理 NL2SQL 服务请求（旧版 API，向后兼容）"""
    return await proxy_request(
        request=request,
        target_base_url=settings.NL2SQL_URL,
        target_path=path,
    )
